server/src/main/java/org/elasticsearch/index/mapper/DocumentFieldMappers.java

```diff
@@ -43,22 +47,54 @@ private static void put(Map<String, Analyzer> analyzers, String key, Analyzer va
     }
 
     public DocumentFieldMappers(Collection<FieldMapper> mappers,
+                                Collection<ObjectMapper> objectMappers,
                                 Collection<FieldAliasMapper> aliasMappers,
                                 Analyzer defaultIndex) {
         Map<String, Mapper> fieldMappers = new HashMap<>();
         Map<String, Analyzer> indexAnalyzers = new HashMap<>();
+        Map<String, ObjectMapper> objects = new HashMap<>();
+
+        boolean hasNested = false;
+        for (ObjectMapper mapper : objectMappers) {
+            if (objects.put(mapper.fullPath(), mapper) != null) {
+                throw new MapperParsingException("Object mapper [" + mapper.fullPath() + "] is defined more than once");
+            }
+            if (mapper.nested().isNested()) {
+                hasNested = true;
+            }
+        }
+        this.hasNested = hasNested;
+
+        int metadataFieldCount = 0;
```

Comment:
Instead of recalculating this, could we pass in the metadata field mappers directly from `DocumentMapper`?

Reply:
I've changed things slightly so that the constructor takes the field count directly, but also added a static factory method that will build a `MappingLookup` from a `Mapper` and `Analyzer`.  This also means that we can move `MapperUtils.collect` into a private static method on `MappingLookup` and remove the `MapperUtils` enum entirely.

Reply:
Great, this turned out nicely.